src/server/cache.py

```diff
@@ -67,19 +71,17 @@ async def allocate_cache(self, descr: TensorDescriptor) -> AsyncContextManager[H
         assert descr.device is None and descr
         allocated_handle = None
         allocated_size_bytes = descr.numel() * torch.finfo(descr.dtype).bits // 8
+        loop = asyncio.get_event_loop()
         try:
-            async with hivemind.utils.enter_asynchronously(self.lock_metadata):
+            async with hivemind.utils.enter_asynchronously(self._lock_acquire_memory):
```

Comment:
**Important discussion:**

Note that due to this lock, only one `_wait_until_available()` can run at a time.

What if someone tries to allocate `max_size_bytes`? Are we sure that all the previous allocations will be eventually freed? If it's not the case, such a query will lead to a deadlock (`MemoryCache` won't allocate neither this large chunks, nor any future smaller chunks).

Even if it's the case, the memory will be underutilized.

Alternatively, we could run `_wait_until_available()` for all the allocations simultaneously - this way, smaller allocations would utilize the recently freed memory. However, if we have a constant stream of small and large mallocs, it may never allow large chunks to be allocated.

Reply:
> Are we sure that all the previous allocations will be eventually freed?
All allocations correspond to currently running rpc_inference queries and leave no data behind.

Reply:
Okay, so the memory is deliberately allocated in FIFO mode.

Good.